akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/Effect.scala

```diff
@@ -58,17 +54,62 @@ object Effect {
  * Not for user extension.
  */
 @DoNotInherit
-trait Effect[+Event, State] { self: EffectImpl[Event, State] ⇒
+trait Effect[+Event, State] {
   /* All events that will be persisted in this effect */
   def events: im.Seq[Event]
 
-  def sideEffects[E >: Event]: im.Seq[ChainableEffect[E, State]]
-
-  /** Convenience method to register a side effect with just a callback function */
+  /**
+   * Run the given callback. Callbacks are run sequentially.
+   */
   final def andThen(callback: State ⇒ Unit): Effect[Event, State] =
-    CompositeEffect(this, SideEffect[Event, State](callback))
+    CompositeEffect(this, ChainedEffect(callback))
+
+  // not named andThen so can pass a pattern match to andThen
+
+  /**
+   *  Run the given callback after the current Effect
+   */
+  final def andChain(chainedEffect: ChainedEffect[State]): Effect[Event, State] =
+    CompositeEffect(this, chainedEffect)
+
+  /**
+   *  Run the given callbacks sequentially after the current Effect
+   */
+  final def andChain(chainedEffects: im.Seq[ChainedEffect[State]]): Effect[Event, State] =
+    CompositeEffect(this, chainedEffects)
 
   /** The side effect is to stop the actor */
-  def andThenStop: Effect[Event, State] =
-    CompositeEffect(this, Effect.stop[Event, State])
+  def andThenStop(): Effect[Event, State] = {
+    CompositeEffect(this, Stop.asInstanceOf[ChainedEffect[State]])
+  }
+}
+
+/**
+ * A ChainedEffect is an side effect that can be chained after a main effect.
```

Comment:
We could call the whole thing side effect but that might start the debate again about our use of effect and side effect terminology. WDYT? It would be:

Effects: Persist, Unhandled, None
Side effects: Callback, Stop

Reply:
I think the separation and names you just proposed make a lot of sense, am I opening pandora's drama box again? I forgot the last time we discussed these... was it from Debasish's feedback?